compatibility layer exceptions are hidden

The compatibility invocation handler doesn't propagate exceptions properly, any invocation exception thrown will get logged and the original TAppplicationException is thrown which is incorrect.

Example stacktrace in the logs:
```19/01/10 09:43:52 WARN closeable.CloseableMetaStoreClientInvocationHandler: Unable to run compatibility for metastore client method getTable. Will rethrow original exception:
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.hotels.hcommon.hive.metastore.client.closeable.CloseableMetaStoreClientInvocationHandler.invokeCompatibility(CloseableMetaStoreClientInvocationHandler.java:66)
	at com.hotels.hcommon.hive.metastore.client.closeable.CloseableMetaStoreClientInvocationHandler.invoke(CloseableMetaStoreClientInvocationHandler.java:53)
	at com.sun.proxy.$Proxy94.getTable(Unknown Source)
	at com.hotels.bdp.circustrain.core.HiveEndpoint.getTable(HiveEndpoint.java:100)
	at com.hotels.bdp.circustrain.core.replica.Replica.validateReplicaTable(Replica.java:291)
	at com.hotels.bdp.circustrain.core.UnpartitionedTableReplication.replicate(UnpartitionedTableReplication.java:84)
	at com.hotels.bdp.circustrain.core.Locomotive.run(Locomotive.java:114)
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:791)
	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:781)
	at org.springframework.boot.SpringApplication.afterRefresh(SpringApplication.java:771)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:307)
	at com.hotels.bdp.circustrain.CircusTrain.main(CircusTrain.java:104)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.hadoop.util.RunJar.run(RunJar.java:234)
	at org.apache.hadoop.util.RunJar.main(RunJar.java:148)
Caused by: NoSuchObjectException(message:db.table1 table not found)
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$get_table_result$get_table_resultStandardScheme.read(ThriftHiveMetastore.java:53203)
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$get_table_result$get_table_resultStandardScheme.read(ThriftHiveMetastore.java:53171)
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$get_table_result.read(ThriftHiveMetastore.java:53102)
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:86)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.hotels.hcommon.hive.metastore.compatibility.HiveMetaStoreClientCompatibility12x.receiveBase(HiveMetaStoreClientCompatibility12x.java:89)
	at com.hotels.hcommon.hive.metastore.compatibility.HiveMetaStoreClientCompatibility12x.recv_get_table(HiveMetaStoreClientCompatibility12x.java:135)
	at com.hotels.hcommon.hive.metastore.compatibility.HiveMetaStoreClientCompatibility12x.get_table(HiveMetaStoreClientCompatibility12x.java:113)
	at com.hotels.hcommon.hive.metastore.compatibility.HiveMetaStoreClientCompatibility12x.getTable(HiveMetaStoreClientCompatibility12x.java:103)
```
The cause `NoSuchObjectException` is not thrown but just logged. It should be thrown.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 * Removed transitive (provided) dependency on `hbase-client`.
 ### Fixed
 * Removed 'isOpen()' method from CloseableMetaStoreClient. See [#11](https://github.com/HotelsDotCom/hcommon-hive-metastore/issues/11).
+* Issue where the wrong exception was being propagated in the compatibility layer.
 
 # [1.2.3] - 2018-10-29
 ### Fixed

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,7 +6,7 @@
 * Removed transitive (provided) dependency on `hbase-client`.
 ### Fixed
 * Removed 'isOpen()' method from CloseableMetaStoreClient. See [#11](https://github.com/HotelsDotCom/hcommon-hive-metastore/issues/11).
-* Issue where the wrong exception was being propagated in the compatibility layer.
+* Issue where the wrong exception was being propagated in the compatibility layer. See [#16](https://github.com/HotelsDotCom/hcommon-hive-metastore/issues/16).
 
 # [1.2.3] - 2018-10-29
 ### Fixed

--- a/README.md
+++ b/README.md
@@ -11,4 +11,4 @@ General purpose libraries for interacting with the Hive metastore
 # Legal
 This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).
 
-Copyright 2018 Expedia Inc.
+Copyright 2018-2019 Expedia Inc.

--- a/src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandler.java
+++ b/src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandler.java
@@ -54,7 +54,7 @@ class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
         if (compatibilityException.getCause().getClass().isAssignableFrom(TApplicationException.class)) {
           log
               .warn(
-                  "Unable to run compatibility for metastore client method {}. Will rethrow original exception, logging exception from compatibility layer: ",
+                  "Invocation of compatibility for metastore client method {} failed. Will rethrow original exception, logging exception from compatibility layer",
                   method.getName(), compatibilityException);
         } else {
           // compatibility worked but threw non TApplicationException, re-throwing cause.
@@ -63,7 +63,7 @@ class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
       } catch (Throwable t) {
         log
             .warn(
-                "Unable to run compatibility for metastore client method {}. Will rethrow original exception, logging exception from compatibility layer: ",
+                "Unable to invoke compatibility for metastore client method {}. Will rethrow original exception, logging exception from invocation handler",
                 method.getName(), t);
       }
       throw delegateException.getCause();

--- a/src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandler.java
+++ b/src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandler.java
@@ -57,7 +57,7 @@ class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
                   "Unable to run compatibility for metastore client method {}. Will rethrow original exception, logging exception from compatibility layer: ",
                   method.getName(), compatibilityException);
         } else {
-          // compatibily worked but threw non TApplicationException, rethrowing cause.
+          // compatibility worked but threw non TApplicationException, re-throwing cause.
           throw compatibilityException.getCause();
         }
       } catch (Throwable t) {

--- a/src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandler.java
+++ b/src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandler.java
@@ -53,12 +53,18 @@ class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
       } catch (InvocationTargetException compatibilityException) {
         if (compatibilityException.getCause().getClass().isAssignableFrom(TApplicationException.class)) {
           log
-              .warn("Unable to run compatibility for metastore client method {}. Will rethrow original exception: ",
+              .warn(
+                  "Unable to run compatibility for metastore client method {}. Will rethrow original exception, logging exception from compatibility layer: ",
                   method.getName(), compatibilityException);
         } else {
-          // compatibily worked but threw non TApplicaitonException, rethrowing cause.
+          // compatibily worked but threw non TApplicationException, rethrowing cause.
           throw compatibilityException.getCause();
         }
+      } catch (Throwable t) {
+        log
+            .warn(
+                "Unable to run compatibility for metastore client method {}. Will rethrow original exception, logging exception from compatibility layer: ",
+                method.getName(), t);
       }
       throw delegateException.getCause();
     }

--- a/src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandler.java
+++ b/src/main/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandler.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2018 Expedia Inc.
+ * Copyright (C) 2018-2019 Expedia Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -13,7 +13,6 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
 package com.hotels.hcommon.hive.metastore.client.closeable;
 
 import java.lang.reflect.InvocationHandler;
@@ -34,9 +33,7 @@ class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
   private final IMetaStoreClient delegate;
   private final HiveMetaStoreClientCompatibility compatibility;
 
-  CloseableMetaStoreClientInvocationHandler(
-      IMetaStoreClient delegate,
-      HiveMetaStoreClientCompatibility compatibility) {
+  CloseableMetaStoreClientInvocationHandler(IMetaStoreClient delegate, HiveMetaStoreClientCompatibility compatibility) {
     this.delegate = delegate;
     this.compatibility = compatibility;
   }
@@ -45,18 +42,25 @@ class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
   public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
     try {
       return method.invoke(delegate, args);
-    } catch (InvocationTargetException e) {
+    } catch (InvocationTargetException delegateException) {
       try {
         log.info("Couldn't invoke method {}", method.toGenericString());
-        if (compatibility != null && e.getCause().getClass().isAssignableFrom(TApplicationException.class)) {
+        if (compatibility != null
+            && delegateException.getCause().getClass().isAssignableFrom(TApplicationException.class)) {
           log.info("Attempting to invoke with {}", compatibility.getClass().getName());
           return invokeCompatibility(method, args);
         }
-      } catch (Throwable t) {
-        log.warn("Unable to run compatibility for metastore client method {}. Will rethrow original exception: ",
-            method.getName(), t);
+      } catch (InvocationTargetException compatibilityException) {
+        if (compatibilityException.getCause().getClass().isAssignableFrom(TApplicationException.class)) {
+          log
+              .warn("Unable to run compatibility for metastore client method {}. Will rethrow original exception: ",
+                  method.getName(), compatibilityException);
+        } else {
+          // compatibily worked but threw non TApplicaitonException, rethrowing cause.
+          throw compatibilityException.getCause();
+        }
       }
-      throw e.getCause();
+      throw delegateException.getCause();
     }
   }
 

--- a/src/main/java/com/hotels/hcommon/hive/metastore/iterator/PartitionIterator.java
+++ b/src/main/java/com/hotels/hcommon/hive/metastore/iterator/PartitionIterator.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2018 Expedia Inc.
+ * Copyright (C) 2018-2019 Expedia Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

--- a/src/test/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandlerTest.java
+++ b/src/test/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandlerTest.java
@@ -135,4 +135,29 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
       assertThat(e, is(original));
     }
   }
+
+  @Test
+  public void invocationExceptionOnCompatibilityLayerIsIgnoredOriginalExceptionShouldBeThrown() throws Throwable {
+    Class<?> clazz = Class.forName(IMetaStoreClient.class.getName());
+    // This method does not exist in compatibility class, will result in NoSuchMethodException from the compatiblity
+    // layer, which should be logged and ignored
+    Method method = clazz.getMethod("getAllDatabases");
+
+    final TApplicationException original = new TApplicationException("original");
+
+    IMetaStoreClient exceptionThrowingClient = Mockito.mock(IMetaStoreClient.class, new Answer<Void>() {
+      @Override
+      public Void answer(InvocationOnMock invocation) throws Throwable {
+        throw original;
+      }
+    });
+
+    invocationHandler = new CloseableMetaStoreClientInvocationHandler(exceptionThrowingClient, compatibility);
+    try {
+      invocationHandler.invoke(null, method, new Object[0]);
+      fail("Exception should have been thrown");
+    } catch (TApplicationException e) {
+      assertThat(e, is(original));
+    }
+  }
 }

--- a/src/test/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandlerTest.java
+++ b/src/test/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandlerTest.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2018 Expedia Inc.
+ * Copyright (C) 2018-2019 Expedia Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -15,12 +15,16 @@
  */
 package com.hotels.hcommon.hive.metastore.client.closeable;
 
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
 import static org.mockito.Matchers.eq;
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
 import java.lang.reflect.Method;
 
 import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
 import org.apache.thrift.TApplicationException;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -55,7 +59,6 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
     String dbName = "db";
     String tableName = "table";
 
-
     IMetaStoreClient exceptionThrowingClient = Mockito.mock(IMetaStoreClient.class, new Answer() {
       @Override
       public Object answer(InvocationOnMock invocation) throws Throwable {
@@ -68,22 +71,66 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
     verify(compatibility).getTable(eq(dbName), eq(tableName));
   }
 
-  @Test(expected = RuntimeException.class)
+  @Test(expected = NoSuchObjectException.class)
   public void dontInvokeCompatibilityWhenExceptionOtherThanTApplicationExceptionIsThrown() throws Throwable {
     Class<?> clazz = Class.forName(IMetaStoreClient.class.getName());
     Method method = clazz.getMethod("getTable", String.class, String.class);
     String dbName = "db";
     String tableName = "table";
 
-
-    IMetaStoreClient exceptionThrowingClient = Mockito.mock(IMetaStoreClient.class, new Answer() {
+    IMetaStoreClient exceptionThrowingClient = Mockito.mock(IMetaStoreClient.class, new Answer<Void>() {
       @Override
-      public Object answer(InvocationOnMock invocation) throws Throwable {
-        throw new RuntimeException();
+      public Void answer(InvocationOnMock invocation) throws Throwable {
+        throw new NoSuchObjectException();
       }
     });
 
     invocationHandler = new CloseableMetaStoreClientInvocationHandler(exceptionThrowingClient, compatibility);
     invocationHandler.invoke(null, method, new String[] { dbName, tableName });
+  }
+
+  @Test(expected = NoSuchObjectException.class)
+  public void rethrowCompatibilityClassException() throws Throwable {
+    Class<?> clazz = Class.forName(IMetaStoreClient.class.getName());
+    Method method = clazz.getMethod("getTable", String.class, String.class);
+    String dbName = "db";
+    String tableName = "table";
+
+    IMetaStoreClient exceptionThrowingClient = Mockito.mock(IMetaStoreClient.class, new Answer<Void>() {
+      @Override
+      public Void answer(InvocationOnMock invocation) throws Throwable {
+        throw new TApplicationException();
+      }
+    });
+
+    when(compatibility.getTable(dbName, tableName)).thenThrow(new NoSuchObjectException("This should be thrown"));
+
+    invocationHandler = new CloseableMetaStoreClientInvocationHandler(exceptionThrowingClient, compatibility);
+    invocationHandler.invoke(null, method, new String[] { dbName, tableName });
+  }
+
+  @Test
+  public void rethrowClientClassExceptionOnTApplicationException() throws Throwable {
+    Class<?> clazz = Class.forName(IMetaStoreClient.class.getName());
+    Method method = clazz.getMethod("getTable", String.class, String.class);
+    String dbName = "db";
+    String tableName = "table";
+
+    final TApplicationException original = new TApplicationException("original");
+    IMetaStoreClient exceptionThrowingClient = Mockito.mock(IMetaStoreClient.class, new Answer<Void>() {
+      @Override
+      public Void answer(InvocationOnMock invocation) throws Throwable {
+        throw original;
+      }
+    });
+
+    when(compatibility.getTable(dbName, tableName)).thenThrow(new TApplicationException("This should not be thrown"));
+
+    invocationHandler = new CloseableMetaStoreClientInvocationHandler(exceptionThrowingClient, compatibility);
+    try {
+      invocationHandler.invoke(null, method, new String[] { dbName, tableName });
+    } catch (TApplicationException e) {
+      assertThat(e, is(original));
+    }
   }
 }

--- a/src/test/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandlerTest.java
+++ b/src/test/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandlerTest.java
@@ -73,7 +73,7 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
   }
 
   @Test(expected = NoSuchObjectException.class)
-  public void dontInvokeCompatibilityWhenExceptionOtherThanTApplicationExceptionIsThrown() throws Throwable {
+  public void delegateThrowsTException() throws Throwable {
     Class<?> clazz = Class.forName(IMetaStoreClient.class.getName());
     Method method = clazz.getMethod("getTable", String.class, String.class);
     String dbName = "db";
@@ -91,7 +91,7 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
   }
 
   @Test(expected = NoSuchObjectException.class)
-  public void rethrowCompatibilityClassException() throws Throwable {
+  public void compatabilityThrowsTException() throws Throwable {
     Class<?> clazz = Class.forName(IMetaStoreClient.class.getName());
     Method method = clazz.getMethod("getTable", String.class, String.class);
     String dbName = "db";
@@ -111,7 +111,7 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
   }
 
   @Test
-  public void rethrowClientClassExceptionOnTApplicationException() throws Throwable {
+  public void delegateAndCompatabilityThrowTApplicationException() throws Throwable {
     Class<?> clazz = Class.forName(IMetaStoreClient.class.getName());
     Method method = clazz.getMethod("getTable", String.class, String.class);
     String dbName = "db";
@@ -139,7 +139,7 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
   @Test
   public void invocationExceptionOnCompatibilityLayerIsIgnoredOriginalExceptionShouldBeThrown() throws Throwable {
     Class<?> clazz = Class.forName(IMetaStoreClient.class.getName());
-    // This method does not exist in compatibility class, will result in NoSuchMethodException from the compatiblity
+    // This method exists in the delegate but not in the compatibility class, will result in NoSuchMethodException from the compatibility
     // layer, which should be logged and ignored
     Method method = clazz.getMethod("getAllDatabases");
 

--- a/src/test/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandlerTest.java
+++ b/src/test/java/com/hotels/hcommon/hive/metastore/client/closeable/CloseableMetaStoreClientInvocationHandlerTest.java
@@ -17,6 +17,7 @@ package com.hotels.hcommon.hive.metastore.client.closeable;
 
 import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
 import static org.mockito.Matchers.eq;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
@@ -59,9 +60,9 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
     String dbName = "db";
     String tableName = "table";
 
-    IMetaStoreClient exceptionThrowingClient = Mockito.mock(IMetaStoreClient.class, new Answer() {
+    IMetaStoreClient exceptionThrowingClient = Mockito.mock(IMetaStoreClient.class, new Answer<Void>() {
       @Override
-      public Object answer(InvocationOnMock invocation) throws Throwable {
+      public Void answer(InvocationOnMock invocation) throws Throwable {
         throw new TApplicationException();
       }
     });
@@ -129,6 +130,7 @@ public class CloseableMetaStoreClientInvocationHandlerTest {
     invocationHandler = new CloseableMetaStoreClientInvocationHandler(exceptionThrowingClient, compatibility);
     try {
       invocationHandler.invoke(null, method, new String[] { dbName, tableName });
+      fail("Exception should have been thrown");
     } catch (TApplicationException e) {
       assertThat(e, is(original));
     }

